src/langs/vala/Compiler.js

```diff
@@ -1,49 +1,79 @@
 import Gio from "gi://Gio";
 import dbus_previewer from "../../Previewer/DBusPreviewer.js";
-import { decode } from "../../util.js";
+import { createTmpDir, copy } from "../../util.js";
+import { setupValaProject } from "./vala.js";
+
+let ready_to_build = false;
+const session_build_dir = await createTmpDir("vala");
 
 export default function ValaCompiler({ session }) {
   const { file } = session;
 
-  const module_file = file.get_child("libworkbenchcode.so");
-  const file_vala = file.get_child("main.vala");
+  const meson_builddir = "builddir";
+  const module_file = session_build_dir
+    .get_child(meson_builddir)
+    .get_child("libworkbenchcode.so");
 
   async function compile() {
-    let args;
+    if (!ready_to_build) {
+      try {
+        await setupValaProject(session_build_dir);
+        ready_to_build = true;
+      } catch (err) {
+        console.error(err);
+        return false;
+      }
+    }
 
-    try {
-      const [contents] = await file_vala.load_contents_async(null);
-      const code = decode(contents);
-      args = getValaCompilerArguments(code);
-    } catch (error) {
-      console.debug(error);
-      return;
+    await copy(
+      "main.vala",
+      file,
+      session_build_dir,
+      Gio.FileCopyFlags.OVERWRITE | Gio.FileCopyFlags.ALL_METADATA,
+    );
+
+    // TODO: Do not run setup if the build directory is already
+    // configured
+    const meson_launcher = new Gio.SubprocessLauncher();
+    meson_launcher.set_cwd(session_build_dir.get_path());
+    const meson_setup = meson_launcher.spawnv([
+      "meson",
+      "setup",
+      meson_builddir,
+    ]);
+
+    await meson_setup.wait_async(null);
+    const setup_result = meson_setup.get_successful();
+    if (!setup_result) {
+      return false;
     }
 
-    const valac_launcher = new Gio.SubprocessLauncher();
-    valac_launcher.set_cwd(file.get_path());
-    const valac = valac_launcher.spawnv([
-      "valac",
-      file_vala.get_path(),
-      "--hide-internal",
-      "-X",
-      "-shared",
-      "-X",
-      "-fpic",
-      "--library",
-      "workbench",
-      "-o",
-      module_file.get_path(),
-      "--vapi",
-      "/dev/null",
-      ...args,
+    const meson_clean = meson_launcher.spawnv([
```

Comment:
When replacing the file in the build dir, meson will not detect any changes on the file, thus not compiling anything and just running the project right away. This is a problem when a user has two open projects and working on them simultaneously, so I went for cleaning the build files and recompiling. I thought this approach was OK since we don't really compile much.